### PR TITLE
better and faster string compression

### PR DESCRIPTION
**Pull request by N-R-K**
instead of using many snprintf + dprintf calls (which are noticeably slow) this patch uses the 8th bit to indicate a run. the decompressor is now faster by avoiding multiple `*printf` calls and the string size has also gone down by about ~200 bytes.

### CODE REVIEW

**Comment by N-R-K** [comment]
```
src/nnn.c:5147:9: warning: ignoring return value of 'write' declared with attribute 'warn_unused_result' [-Wunused-result]
 5147 |         write(fd, g_buf, w - g_buf);
```

Cringe. None of the `dprintf` calls were checked either, casting the result to `void` also doesn't work. Had to add a dummy `if` around the `write` call.

I'll take a look into the failing patch in a bit.

**Review by KlzXS — APPROVED** [review]
I'll trust you that the patches work as they should now.

While I don't like these magic octal spacers as they're hard to read at a glance and would much rather we use some macro for inserting them, it seems that there is no good way to do that. Maybe if we add a header with a bunch of different spacing macros defined it could be done. Something like `SPACER_10` or `SPACER(10)` which expands to `SPACER_10`.  But alas we don't change this often, do we?

The decoding works just fine and produces what you would expect.

PS. Next time please tag me with @KlzXS. That's the main thing I look for when going through emails. I almost missed this.

**Comment by N-R-K** [comment]
> would much rather we use some macro for inserting them, it seems that there is no good way to do that.

Yup, that's what I was trying to do at first too. But didn't find any simple way to do it.

**Comment by N-R-K** [comment]
My main problem was the multiple `*printf`s, they're noticeably slow.

I can try to revert back to old-ish string, but I'd like to remove those printf calls with something faster.

**Comment by N-R-K** [comment]
Also, I thought about using a "generator" similar to the icon hash-table. But it seemed a bit too overkill for something like this, so I didn't pursue it.

**Comment by KlzXS** [comment]
> I find it extremely difficult to decipher.

We could still go for a header file that has something like this:
```c
#define SPACER_0 "\200" // 0x80 == 0o200 == 128
#define SPACER_1 "\201" // perhaps these two are useless, but for the sake of clarity
...
#define SPACER_32 "\240" // 32 spaces ought to be enough for anyone
```
or something similar. That's plenty readable, maybe even more than the current format, but it would require an extra file (I assume none of us is keen on just pasting something like that into `nnn.c` or `nnn.h` directly.

A generator crossed my mind as well, though I wouldn't call it overkill as much as way too complicated for the given task.

@N-R-K just to be clear about the benefits of this approach. You say the string is down ~200 bytes, I assume the binary size has at least as much saved. You also say "which are noticeably slow" and that "the decompressor is now faster". What does "noticeably slow" and "faster" mean in this context? I can't image we are talking about more than a few milliseconds here.

**Comment by N-R-K** [comment]
@KlzXS The main reason I was looking into this is because everytime I press `?` it takes about half a second for the help page to show up the first time (after the first time, it's much faster).

The printf calls were contributing to that, but only a couple miliseconds, like you said. The main culprit seems to be the `spawn` call. It might be something related to my system - I'll need to investigate that (for context, I'm using the `lesspipe` script, I think - which might have something to do with this).

But in any case, I think it shouldn't be too difficult to do better than double printf calls just to convert some special chars to spaces.

**Comment by N-R-K** [comment]
@jarun How does this look? It's the same before except `%-14c` is replaced with just `%14` and is decoded without any printfs.

```diff
diff --git a/src/nnn.c b/src/nnn.c
index 592a083a..a20fa43b 100644
--- a/src/nnn.c
+++ b/src/nnn.c
@@ -5092,42 +5092,42 @@ static void show_help(const char *path)
 	const char helpstr[] = {
 	"0\n"
 	"1NAVIGATION\n"
-	       "9Up k  Up%-16cPgUp ^U  Page up\n"
-	       "9Dn j  Down%-14cPgDn ^D  Page down\n"
-	       "9Lt h  Parent%-12c~ ` @ -  ~, /, start, prev\n"
-	   "5Ret Rt l  Open%-20c'  First file/match\n"
-	       "9g ^A  Top%-21cJ  Jump to entry/offset\n"
-	       "9G ^E  End%-20c^J  Toggle auto-advance on open\n"
-	      "8B (,)  Book(mark)%-11cb ^/  Select bookmark\n"
-		"a1-4  Context%-11c(Sh)Tab  Cycle/new context\n"
-	    "62Esc ^Q  Quit%-20cq  Quit context\n"
-		 "b^G  QuitCD%-18cQ  Pick/err, quit\n"
+	       "9Up k  Up%16PgUp ^U  Page up\n"
+	       "9Dn j  Down%14PgDn ^D  Page down\n"
+	       "9Lt h  Parent%12~ ` @ -  ~, /, start, prev\n"
+	   "5Ret Rt l  Open%20'  First file/match\n"
+	       "9g ^A  Top%21J  Jump to entry/offset\n"
+	       "9G ^E  End%20^J  Toggle auto-advance on open\n"
+	      "8B (,)  Book(mark)%11b ^/  Select bookmark\n"
+		"a1-4  Context%11(Sh)Tab  Cycle/new context\n"
+	    "62Esc ^Q  Quit%20q  Quit context\n"
+		 "b^G  QuitCD%18Q  Pick/err, quit\n"
 	"0\n"
 	"1FILTER & PROMPT\n"
-		  "c/  Filter%-17c^N  Toggle type-to-nav\n"
-		"aEsc  Exit prompt%-12c^L  Toggle last filter\n"
-		  "c.  Toggle hidden%-5cAlt+Esc  Unfilter, quit context\n"
+		  "c/  Filter%17^N  Toggle type-to-nav\n"
+		"aEsc  Exit prompt%12^L  Toggle last filter\n"
+		  "c.  Toggle hidden%05Alt+Esc  Unfilter, quit context\n"
 	"0\n"
 	"1FILES\n"
-	       "9o ^O  Open with%-15cn  Create new/link\n"
-	       "9f ^F  File stats%-14cd  Detail mode toggle\n"
-		 "b^R  Rename/dup%-14cr  Batch rename\n"
-		  "cz  Archive%-17ce  Edit file\n"
-		  "c*  Toggle exe%-14c>  Export list\n"
-	    "6Space +  (Un)select%-12cm-m  Select range/clear\n"
-	          "ca  Select all%-14cA  Invert sel\n"
-	       "9p ^P  Copy here%-12cw ^W  Cp/mv sel as\n"
-	       "9v ^V  Move here%-15cE  Edit sel list\n"
-	       "9x ^X  Delete%-18cS  Listed sel size\n"
+	       "9o ^O  Open with%15n  Create new/link\n"
+	       "9f ^F  File stats%14d  Detail mode toggle\n"
+		 "b^R  Rename/dup%14r  Batch rename\n"
+		  "cz  Archive%17e  Edit file\n"
+		  "c*  Toggle exe%14>  Export list\n"
+	    "6Space +  (Un)select%12m-m  Select range/clear\n"
+	          "ca  Select all%14A  Invert sel\n"
+	       "9p ^P  Copy here%12w ^W  Cp/mv sel as\n"
+	       "9v ^V  Move here%15E  Edit sel list\n"
+	       "9x ^X  Delete%18S  Listed sel size\n"
 		"aEsc  Send to FIFO\n"
 	"0\n"
 	"1MISC\n"
-	      "8Alt ;  Select plugin%-11c=  Launch app\n"
-	       "9! ^]  Shell%-19c]  Cmd prompt\n"
-		  "cc  Connect remote%-10cu  Unmount remote/archive\n"
-	       "9t ^T  Sort toggles%-12cs  Manage session\n"
-		  "cT  Set time type%-11c0  Lock\n"
-		 "b^L  Redraw%-18c?  Help, conf\n"
+	      "8Alt ;  Select plugin%11=  Launch app\n"
+	       "9! ^]  Shell%19]  Cmd prompt\n"
+		  "cc  Connect remote%10u  Unmount remote/archive\n"
+	       "9t ^T  Sort toggles%12s  Manage session\n"
+		  "cT  Set time type%110  Lock\n"
+		 "b^L  Redraw%18?  Help, conf\n"
 	};
 
 	int fd = create_tmp_file();
@@ -5144,17 +5144,23 @@ static void show_help(const char *path)
 	if (prog)
 		get_output(prog, NULL, NULL, fd, FALSE);
 
-	start = end = helpstr;
-	while (*end) {
-		if (*end == '\n') {
-			snprintf(g_buf, CMD_LEN_MAX, "%*c%.*s",
-				 xchartohex(*start), ' ', (int)(end - start), start + 1);
-			dprintf(fd, g_buf, ' ');
-			start = end + 1;
+	bool hex = true;
+	char *w = g_buf;
+	end = helpstr + (sizeof helpstr - 1);
+	for (start = helpstr; start < end; ++start) {
+		if (hex) {
+			for (int k = 0, n = xchartohex(*start); k < n; ++k) *w++ = ' ';
+			hex = false;
+		} else if (*start == '%') {
+			int n = ((start[1] - '0') * 10) + (start[2] - '0');
+			for (int k = 0; k < n; ++k) *w++ = ' ';
+			start += 2;
+		} else {
+			*w++ = *start;
+			hex = *start == '\n';
 		}
-
-		++end;
 	}
+	write(fd, g_buf, w - g_buf);
 
 	dprintf(fd, "\nLOCATIONS\n");
 	for (uchar_t i = 0; i < CTX_MAX; ++i)
```

**Comment by jarun** [comment]
Yes this is much better. @KlzXS what do you think?

**Comment by KlzXS** [comment]
That's better. Also allows for the use of Unicode characters which is good. Doesn't save as many characters, but oh well. 

Just document that `%XX` spacing __*must*__ be two digits as per the rest of your code.
`"cc  Connect remote%10u  Unmount remote/archive\n"` if this line ever changes and we require 9 spaces it should be `%09` otherwise things will break. I don't think that's likely, but if someone translates `nnn` into their preferred language it could happen.

Also don't forget the horse. There really is no reason not to include it in `helpstr`.

File: src/nnn.c [hunk 1]
```diff
@@ -5079,82 +5079,87 @@
...
+	bool hex = true;
+	char *w = g_buf;
+	const char *end = helpstr + (sizeof helpstr - 1);
+	for (const char *s = helpstr; s < end; ++s) {
+		if (hex) {
+			for (int k = 0, n = xchartohex(*s); k < n; ++k) *w++ = ' ';
+		} else if (*s == '%') {
+			int n = ((s[1] - '0') * 10) + (s[2] - '0');
+			for (int k = 0; k < n; ++k) *w++ = ' ';
+			s += 2;
```
[the remark this answers is not in the document]
Comment (jarun) [reply src/nnn.c:5156]: @N-R-K please check.

[the remark this answers is not in the document]
Comment (N-R-K) [reply src/nnn.c:5156]: The code is fine. It's nagging about stylistic issues. I can try to silence the warning but it'll just make the source bigger for no real reasons.

**Comment by jarun** [comment]
We'll ignore it then. Just close it with some comments.

**Comment by jarun** [comment]
Thank you!

**Pull request merged**